modules/tensor_mechanics/src/materials/ADViscoplasticityStressUpdateBase.C

```diff
@@ -0,0 +1,130 @@
+                                          "porosity",
+                                          "Name of porosity material property");
+    params.addParam<std::string>("total_strain_base_name", "Base name for the total strain");
+
+    params.addParamNamesToGroup("inelastic_strain_name", "Advanced"););
+
+template <ComputeStage compute_stage>
+ADViscoplasticityStressUpdateBase<compute_stage>::ADViscoplasticityStressUpdateBase(
+    const InputParameters & parameters)
+  : ADStressUpdateBase<compute_stage>(parameters),
+    ADSingleVariableReturnMappingSolution<compute_stage>(parameters),
+    _total_strain_base_name(isParamValid("total_strain_base_name")
+                                ? getParam<std::string>("total_strain_base_name") + "_"
+                                : ""),
+    _strain_increment(
+        getADMaterialProperty<RankTwoTensor>(_total_strain_base_name + "strain_increment")),
+    _effective_inelastic_strain(declareADProperty<Real>(
+        _base_name + "effective_" + getParam<std::string>("inelastic_strain_name"))),
+    _effective_inelastic_strain_old(getMaterialPropertyOld<Real>(
+        _base_name + "effective_" + getParam<std::string>("inelastic_strain_name"))),
+    _inelastic_strain(declareADProperty<RankTwoTensor>(
+        _base_name + getParam<std::string>("inelastic_strain_name"))),
+    _inelastic_strain_old(getMaterialPropertyOld<RankTwoTensor>(
+        _base_name + getParam<std::string>("inelastic_strain_name"))),
+    _max_inelastic_increment(getParam<Real>("max_inelastic_increment")),
+    _intermediate_porosity(0.0),
+    _porosity_old(getMaterialPropertyOld<Real>(getParam<MaterialPropertyName>("porosity_name"))),
+    _verbose(getParam<bool>("verbose")),
+    _derivative(0.0)
+{
+}
+
+template <ComputeStage compute_stage>
+void
+ADViscoplasticityStressUpdateBase<compute_stage>::initQpStatefulProperties()
+{
+  _effective_inelastic_strain[_qp] = 0.0;
+  _inelastic_strain[_qp].zero();
+}
+
+template <ComputeStage compute_stage>
+void
+ADViscoplasticityStressUpdateBase<compute_stage>::propagateQpStatefulProperties()
+{
+  _effective_inelastic_strain[_qp] = _effective_inelastic_strain_old[_qp];
+  _inelastic_strain[_qp] = _inelastic_strain_old[_qp];
+}
+
+template <ComputeStage compute_stage>
+Real
+ADViscoplasticityStressUpdateBase<compute_stage>::computeReferenceResidual(
+    const ADReal & /*effective_trial_stress*/, const ADReal & gauge_stress)
+{
+  // Use gauge stress for relative tolerance criteria, defined as:
+  // std::abs(residual / gauge_stress) <= _relative_tolerance
+  return MetaPhysicL::raw_value(gauge_stress);
+}
+
+template <ComputeStage compute_stage>
+Real
+ADViscoplasticityStressUpdateBase<compute_stage>::computeTimeStepLimit()
+{
+  const Real scalar_inelastic_strain_incr =
+      MetaPhysicL::raw_value(_effective_inelastic_strain[_qp]) -
+      _effective_inelastic_strain_old[_qp];
+
+  if (MooseUtils::absoluteFuzzyEqual(scalar_inelastic_strain_incr, 0.0))
```

Comment:
:-)   you love these absoluteFuzzyEqual, heh?!  Do you actually just want ==0 here?

Reply:
Yeah, copied from previous implementations, it's probably just as good to do ==0 here